src/backend/mod.rs

```diff
@@ -0,0 +1,222 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! This module defines the data structures used for the intermmediate representation (IR),
+//! as well as the logic for compiling the filter into BPF code, the final form of the filter.
+
+mod bpf;
+mod condition;
+mod filter;
+mod rule;
+
+pub use condition::SeccompCondition;
+pub use filter::SeccompFilter;
+pub use rule::SeccompRule;
+
+use core::fmt::Formatter;
+use std::convert::TryFrom;
+use std::fmt::Display;
+
+use bpf::{
+    ARG_NUMBER_MAX, AUDIT_ARCH_AARCH64, AUDIT_ARCH_X86_64, BPF_MAX_LEN, SECCOMP_RET_ALLOW,
+    SECCOMP_RET_ERRNO, SECCOMP_RET_KILL_PROCESS, SECCOMP_RET_KILL_THREAD, SECCOMP_RET_LOG,
+    SECCOMP_RET_MASK, SECCOMP_RET_TRACE, SECCOMP_RET_TRAP,
+};
+
+pub use bpf::{sock_filter, BpfProgram, BpfProgramRef};
+
+/// Backend Result type.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Backend-related errors.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// Attempting to associate an empty vector of conditions to a rule.
+    EmptyRule,
+    /// Filter exceeds the maximum number of instructions that a BPF program can have.
+    FilterTooLarge(usize),
+    /// Filter and default actions are equal.
+    IdenticalActions,
+    /// Argument index of a `SeccompCondition` exceeds the maximum linux syscall index.
+    InvalidArgumentNumber,
+    /// Invalid TargetArch.
+    InvalidTargetArch(String),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        use self::Error::*;
+
+        match self {
+            EmptyRule => {
+                write!(f, "The condition vector of a rule cannot be empty.")
+            }
+            FilterTooLarge(len) => write!(
+                f,
+                "The seccomp filter contains too many BPF instructions: {}. Max length is {}.",
+                len, BPF_MAX_LEN
+            ),
+            IdenticalActions => write!(f, "`filter_action` and `default_action` are equal."),
+            InvalidArgumentNumber => {
+                write!(
+                    f,
+                    "The seccomp rule contains an invalid argument index. Maximum index value: {}",
+                    ARG_NUMBER_MAX
+                )
+            }
+            InvalidTargetArch(arch) => write!(f, "Invalid target arch: {}.", arch.to_string()),
+        }
+    }
+}
+
+/// Supported target architectures.
+#[allow(non_camel_case_types)]
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum TargetArch {
+    /// x86_64 arch
+    x86_64,
+    /// aarch64 arch
+    aarch64,
+}
+
+impl TargetArch {
+    /// Get the arch audit value. Used for the runtime arch check embedded in the BPF filter.
+    fn get_audit_value(self) -> u32 {
```

Comment:
This function is not needed because enum values are numeric types. We can instead define `TargetArch` as follows:

```rust
#[allow(non_camel_case_types)]
#[derive(Debug, PartialEq, Clone, Copy)]
pub enum TargetArch {
    /// x86_64 arch
    x86_64 = AUDIT_ARCH_X86_64 as isize,
    /// aarch64 arch
    aarch64 = AUDIT_ARCH_AARCH64 as isize,
}
```

And then use its value directly (or convert it to other numeric types if needed).